### PR TITLE
Added Support for 3DYMY's BT7272A version of the RAMPS1.4Plus Board

### PR DESCRIPTION
**Pull request by RobAltenburg**
The BT7272A has a different pinout from the standard RAMPS_EHB.  Also, their 2004LCD has a different pinout than the REPRAP_DISCOUNT_SMARTCONTROLLER it clones.

### CODE REVIEW

**Comment by hominidae** [comment]
I've submitted an additional pull request to your copy. I found and fixed a small typo and made sure that pins.h and boards.h pointed at each other correctly.

I have it running on my board now. The 3DYMY boards have nearly no documentation, and with the pins changed, the display doesn't work out of the box. Thanks a bunch!

**Comment by iain1965** [comment]
when is this board going to be added to the release candidate?
I have checked the latest release boards.h file and still not included.

**Comment by hominidae** [comment]
I asked, no reply yet. Sounds like they're busy with other stuff.

**Comment by jackshi618** [comment]
I wrote about the definition of 3dYMY

#ifndef __AVR_ATmega1280__
 #ifndef __AVR_ATmega2560__
 #error Oops!  Make sure you have 'Arduino Mega' selected from the 'Tools -> Boards' menu.
 #endif
#endif

#ifndef BOARD_NAME
  #define BOARD_NAME "3DYMY"
#endif

    #define X_STEP_PIN         54
    #define X_DIR_PIN          55
    #define X_ENABLE_PIN       38
    #define X_MIN_PIN           3
    #define X_MAX_PIN           2

    #define Y_STEP_PIN         60
    #define Y_DIR_PIN          61
    #define Y_ENABLE_PIN       56
    #define Y_MIN_PIN          14
    #define Y_MAX_PIN          15

    #define Z_STEP_PIN         46
    #define Z_DIR_PIN          48
    #define Z_ENABLE_PIN       62
    #define Z_MIN_PIN          18
    #define Z_MAX_PIN          19

    #define Z2_STEP_PIN         26
    #define Z2_DIR_PIN          28
    #define Z2_ENABLE_PIN       24

    #define E0_STEP_PIN        36 //3dymy
    #define E0_DIR_PIN         34 //3dymy
    #define E0_ENABLE_PIN      30 //3dymy

    #define E1_STEP_PIN        26 //3dymy
    #define E1_DIR_PIN         28 //3dymy
    #define E1_ENABLE_PIN      24 //3dymy

    #define SDPOWER            -1
    #define SDSS               53
    #define LED_PIN            13

    #define FAN_PIN             9  //(Sprinter config)

    #define PS_ON_PIN          12
    
    #define KILL_PIN           27

    #define HEATER_0_PIN        8
    #define HEATER_1_PIN       -1
    #define HEATER_2_PIN       -1

    #define TEMP_0_PIN         13    // MUST USE ANALOG INPUT NUMBERING NOT DIGITAL OUTPUT NUMBERING!!!!!!!!!
    #define TEMP_1_PIN         15
    #define TEMP_2_PIN         -1
    #define HEATER_BED_PIN     10
    
    #define TEMP_BED_PIN       14    // MUST USE ANALOG INPUT NUMBERING NOT DIGITAL OUTPUT NUMBERING!!!!!!!!!

    #ifdef NUM_SERVOS
      #define SERVO0_PIN       11
    #endif

    #ifdef ULTRA_LCD
      #ifdef REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
        #define encrot0     0
        #define encrot1     2
        #define encrot2     3
        #define encrot3     1
        #define BLEN_A      0
        #define BLEN_B      1
        #define BLEN_B      2

        #define LCD_PINS_RS      42
        #define LCD_PINS_ENABLE  44
        #define LCD_PINS_D4      23    
        #define LCD_PINS_D5      25
        #define LCD_PINS_D6      27
        #define LCD_PINS_D7      53

        #define BEEPER_PIN   37
        #define BTN_EN1  31
        #define BTN_EN2  33
        #define BTN_ENC  35
        #define SD_DETECT_PIN  49
      #elif defined(NEWPANEL)
      //#else  //LCD2004
        #define LCD_PINS_RS        41
        #define LCD_PINS_ENABLE    31
        #define LCD_PINS_D4        33
        #define LCD_PINS_D5        23
        #define LCD_PINS_D6        42
        #define LCD_PINS_D7        44
        #define BEEPER_PIN     37
        #define BTN_EN1    25
        #define BTN_EN2    29
        #define BTN_ENC    35
        #define SD_DETECT_PIN 49
      #endif
    #endif  //ULTRA_LCD

**Comment by Roxy-3D** [comment]
First....   This needs to be squashed.   There is no way to review it.

But why is this RAMPS board a good thing?  It appears to be a non-standard one off that doesn't even try to be compatible with what has gone before it?

The reason I'm asking is every board we add to the examples folder results in more work as we add (or edit) options.    If only a few people are using this board...   It makes more sense to just point those people to this Pull Request.

**Comment by fiveangle** [comment]
While looking for a good RAMPS1.4, I came across this "RAMPS1.4plus" and after my research, this is  pretty much in the same category as the "Prusa i4" tactics.  There's nothing "plus" about,  and it was named in an attempt to  unscrupulously confuse and ride on the coattails of RAMPS1.4 popularity.

**Comment by iain1965** [comment]
seriously who cares?

**Comment by iain1965** [comment]
secondly I've unsubscribed from this thread 3 times now and keep getting notifications, please sort this out.

**Comment by fiveangle** [comment]
because it can confuse people who are trying to configure a new printer and thinking they are related

**Comment by hominidae** [comment]
There are at least 150 or so folks with this specific board. [FLSUN Prusa-clone Thingiverse group](https://www.thingiverse.com/groups/flsun-prusa-user-group)

As indicated earlier, there are a number of issues that stem from the Chinese suppliers of these boards (and 3D printers) having changed the pin configuration of a standard Ramps 1.4 board.

The core issue is that by adding a board and pins configuration that abstracts users from having to change those files themselves means that the Marlin firmware can be compiled easily by end users by simply taking advantage of the pre-existing system within the Marlin firmware used to make the Marlin firmware more compatible with different environments.

I'm sure there are a number of people willing to test, verify and confirm that new releases work as intended on these cheap Chinese Ramps 1.4 clones. Myself included.

The changes listed here, along with the pull request I submitted to RobAltenburg's repository are confirmed to work with my printer at least. [Prior to that, the code RobAltenburg submitted wouldn't compile](https://github.com/RobAltenburg/Marlin/pull/1/commits/d9ee81003ade9eeb72344087f99d46832abf0996) (as the changes incorrectly referenced each other); and prior to that, the Marlin firmware could be pushed to the board and result in a 3D printer without a display and that didn't work.

I spent a good bit of time troubleshooting just that until I got the latest Marlin firmware working using RobAltenburg's pull request.

In addition, by solving this with a boards and pins declaration it saves people asking questions everywhere which requires time spent doing complicated troubleshooting thirty or forty different times each time when the answer is simple and obvious.

[Obligatory example from the Reprap forum](http://forums.reprap.org/read.php?146,759185)

**Comment by fiveangle** [comment]
the question isn't really about adding BOARD_BT7272A, the question is the example configuration really needed ?  as long as the board defnition is there, they can find the appropriate configuration*.h within the community, right ?  the reason is, for a relative handful of boards, it's a pain to have to maintain the configuration examples for every change in perpetuity that is made to configuration*.h.  plus someone configuring Marlin for their new printer may think the configuration example is relevant to their RAMPS (which it amost surely isn't).   can we omit the example configuration ?

and again, needs to be squashed

**Comment by hominidae** [comment]
Ah, now I get you.

Agreed. The BT7272A folder within the "example_configurations/" directory is likely unnecessary.

However, if the "pins.h" and the "pins_BT7272A.h" file had an extended readme within the "pins_BT7272A.h" file itself explaining how to call it would likely save a lot of folks the trouble.

That would simplify these changes to just a single new file and a single addition to "pins.h" that could be called by users if needed without needing the superfluous example configuration.

The files within the BT7272A folder are all unnecessary except as an example for a working configuration, and would likely need to be changed by users anyway for whatever LCD, thermistor, and other equipment they'd be using anyway.

**Comment by fiveangle** [comment]
Yes, you can think of the pins_BT7272A.h file as the BT272A community's own little haven that you have carte blanche to outfit as you collectively see fit in order to best serve owners of those boards.  Outside of that, changes are more scrutinised as to not make things difficult or confusing for owners of other boards or new users.

Agree with @hominidae that drop all changes except those needed to implemnt the pins_BT7272A.h file, squash, and wait-for-review/merge (it may be a bit as Marlin is currently in somewhat of a code freeze while the 2.0 project layout is being done.  Once you've done that, if it takes a while to get looked at, since this is a simple change, Scott will likely file a separate PR with the pins*.h changes and close this one, rather than asking for subsequent squash/rebase).

File: Marlin/pins_BT7272A.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#ifndef BOARD_NAME
+  #define BOARD_NAME "BT7272A"
+#endif
+
+#ifndef DEFAULT_MACHINE_NAME
+  #define DEFAULT_MACHINE_NAME "flsun i3"
+#endif
+
+#ifndef DEFAULT_SOURCE_CODE_URL
+  #define DEFAULT_SOURCE_CODE_URL "https://github.com/RobAltenburg/Marlin"
```
Comment (fiveangle) [inline Marlin/pins_BT7272A.h]: i believe this should be removed by default

**Comment by fiveangle** [comment]
Forgot to mention: this PR is opened against the "RC" branch but it needs to target either bugfix1.1.x (okay) or bugfix2.0.x (ideal).  Sorry I didn't notice that before :(

**Pull request closed without merging**